plugins/bigquery/dbt/adapters/bigquery/impl.py

```diff
@@ -430,17 +431,73 @@ def is_replaceable(self, relation, conf_partition, conf_cluster):
         except google.cloud.exceptions.NotFound:
             return True
 
-        table_partition = table.time_partitioning
-        if table_partition is not None:
-            table_partition = table_partition.field
-
         table_cluster = table.clustering_fields
 
         if isinstance(conf_cluster, str):
             conf_cluster = [conf_cluster]
 
-        return table_partition == conf_partition \
-            and table_cluster == conf_cluster
+        time_partition = table.time_partitioning
+        range_partition = table.range_partitioning
+
+        if time_partition is not None:
+            return time_partition.field == conf_partition.get('field', None) \
+                and table_cluster == conf_cluster
+
+        if range_partition is not None:
+            table_range = range_partition.range_
+            conf_range = conf_partition.get('range', None)
+            return range_partition.field == conf_partition.get('field', None) \
+                and table_range.start == conf_range.get('start', None) \
+                and table_range.end == conf_range.get('end', None) \
+                and table_range.interval == conf_range.get('interval', None) \
+                and table_cluster == conf_cluster
+
+    @available
+    def parse_partition_by(self, raw_partition_by):
+        """
+        dbt v0.16.0 expects `partition_by` to be a dictionary where previously
+        it was a string. Check the type of `partition_by`, raise error
+        or warning if string, and attempt to convert to dict.
+        """
+        if isinstance(raw_partition_by, dict):
+            if raw_partition_by.get('field', None) is not None:
+                if raw_partition_by.get('data_type', None) is not None:
+                    return raw_partition_by
+                else:  # assume date type as default
+                    return dict(raw_partition_by, data_type='date')
+            else:
+                dbt.exceptions.CompilerException(
+                    'Config `partition_by` is missing required item `field`'
+                )
+
+        elif isinstance(raw_partition_by, str):
+            if 'range_bucket' in raw_partition_by.lower():
+                dbt.exceptions.CompilerException('''
+                    BigQuery integer range partitioning (beta) is supported \
+                    by the new `partition_by` config, which accepts a \
+                    dictionary. See: [dbt docs link TK]
+                    ''')  # TODO
+            else:
+                p = re.compile(
```

Comment:
This regex is a little too complicated for my liking... can we try something a little simpler instead? How about we try something like:

```py
partition_by = raw_partition_by.strip()
if partition_by.lower().startswith('date('):
  partition_by = re.match('date\((.*)\)', partition_by, re.IGNORECASE)
  data_type = 'date'
else:
  data_type = 'timestamp'

inferred_partition_by = ...
```

Let me know if I'm missing something important in the logic here!